ssz_serialization/codec.nim

```diff
@@ -277,14 +291,21 @@ proc readSszValue*[T](input: openArray[byte],
     if resultBytesCount == maxExpectedSize:
       checkForForbiddenBits(T, input, val.maxLen + 1)
 
-  elif val is HashList | HashArray:
+  elif val is HashList:
+    readSszValue(input, val.data, hashListClosures(addr val))
+  elif val is HashArray:
     readSszValue(input, val.data)
     val.resetCache()
   elif val is Digest:
     readSszValue(input, val.data)
   elif val is List|array:
     type E = ElemType(type val)
 
+    # This is arbitrary, but keeps the initial implementation from affecting
```

Comment:
that said, we can apply the same optimisation to arrays (ie on short reorgs, the arrays mostly stay the same to

Reply:
Yeah, part of this was that the initial version has so much code duplication. If it has less, that's more feasible.

The other aspect is that the element-wise comparisons have some overhead, and I want these to be justified by benchmark wins (see the with/without fixed-size benchmarks; I was ready to remove those too for a first implementation, but it's surprisingly compelling).

Reply:
The arrays also don't grow in overall size. The validators are getting actively slower over time, but the arrays aren't.

Reply:
> I want these to be justified by benchmark wins

well, now that the lists are "better" we'll start seeing more of the array as a percentage of total time - ie we have the 64k-hashes arrays that between one epoch and the other change only 32 values - that saves something along the lines of 63k*log2(63k) calls to the hashing library which can't be nothing

Reply:
Yeah, agree, it's a good next improvement, to at least benchmark, and of the remaining performance improvements probably one of the more localized/straightforward ones.